### PR TITLE
Add Sve.ScatterXBitYNarrowing() on Arm64

### PR DESCRIPTION
**Pull request by SwapnilGaikwad**
Contribute towards https://github.com/dotnet/runtime/issues/99957.

Includes:
- Sve.Scatter16BitNarrowing()
- Sve.Scatter8BitNarrowing()
- Sve.Scatter32BitNarrowing()
- Sve.Scatter16BitWithByteOffsetsNarrowing()
- Sve.Scatter8BitWithByteOffsetsNarrowing()
- Sve.Scatter32BitWithByteOffsetsNarrowing()

@a74nh @kunalspathak @dotnet/arm64-contrib @arch-arm64-sve

### CODE REVIEW

**Comment by SwapnilGaikwad** [comment]
<details>
<summary>All scatter store stress tests are passing</summary>

```
===================Running default===================
------------------- {} -------------------
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_bases_double_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_bases_long_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_bases_ulong_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_double_long() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_double_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_int_int() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_int_uint() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_long_long() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_long_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_float_int() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_float_uint() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_uint_int() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_uint_uint() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_ulong_long() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter_indices_ulong_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter16BitNarrowing_long_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter16BitNarrowing_ulong_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter16BitWithByteOffset_int_int() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter16BitWithByteOffset_int_uint() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter16BitWithByteOffset_uint_int() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter16BitWithByteOffset_uint_uint() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter16BitWithByteOffset_long_long() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter16BitWithByteOffset_long_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter16BitWithByteOffset_ulong_long() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter16BitWithByteOffset_ulong_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter32BitNarrowing_long_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter32BitNarrowing_ulong_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter32BitWithByteOffset_long_long() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter32BitWithByteOffset_long_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter32BitWithByteOffset_ulong_long() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter32BitWithByteOffset_ulong_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter8BitNarrowing_long_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter8BitNarrowing_ulong_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter8BitWithByteOffset_int_int() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter8BitWithByteOffset_int_uint() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter8BitWithByteOffset_uint_int() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter8BitWithByteOffset_uint_uint() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter8BitWithByteOffset_long_long() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter8BitWithByteOffset_long_ulong() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter8BitWithByteOffset_ulong_long() : 9
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve._Sve.Program.Sve_Scatter8BitWithByteOffset_ulong_ulong() : 9
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------
```
</details>

**Review by TIHan — COMMENTED** [review]
So far, this look good. Will give another look when it's ready for review.

**Comment by SwapnilGaikwad** [comment]
> So far, this look good. Will give another look when it's ready for review.

Thanks @TIHan. I had kept it as a draft because I wasn't sure about how we want to deal with the conflicting APIs that are currently commented 🙂 .

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.cs [hunk 1]
```diff
@@ -7303,6 +7303,280 @@ public new abstract class Arm64 : AdvSimd.Arm64
...
+        // void svst1h_scatter_[u32]offset[_s32](svbool_t pg, int16_t *base, svuint32_t offsets, svint32_t data)
+        //   ST1H Zdata.S, Pg, [Xbase, Zoffsets.S, UXTW]
+        // </summary>
+        public static unsafe void Scatter16BitWithByteOffsetsNarrowing(Vector<int> mask, short* address, Vector<uint> offsets, Vector<int> data) => Scatter16BitWithByteOffsetsNarrowing(mask, address, offsets, data);
+
+        // <summary>
+        // void svst1h_scatter_[s32]index[_s32](svbool_t pg, int16_t *base, svint32_t indices, svint32_t data)
+        //   ST1H Zdata.S, Pg, [Xbase, Zindices.S, SXTW #1]
+        // </summary>
+        // public static unsafe void Scatter16BitWithByteOffsetsNarrowing(Vector<int> mask, short* address, Vector<int> indices, Vector<int> data) => Scatter16BitWithByteOffsetsNarrowing(mask, address, indices, data);
```
Comment (kunalspathak) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.cs:7353]: as discussed offline, please remove the commented methods from this and other files.

**Review by kunalspathak — COMMENTED** [review]
Looks good overall. Added some nit comments.

**Review by kunalspathak — APPROVED** [review]
LGTM. Thanks!

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs [hunk 2]
```diff
@@ -4031,6 +4031,53 @@
...
 
+
+    ("SveScatterVectorBases.template",         new Dictionary<string, string> { ["TestName"] = "Sve_Scatter_bases_double_ulong",                                                                   ["Isa"] = "Sve",           ["LoadIsa"] = "Sve",     ["Method"] = "Scatter",                                                                                                                            ["Op1VectorType"] = "Vector",    ["Op1BaseType"] = "Double", ["Op2VectorType"] = "Vector",    ["Op2BaseType"] = "UInt64",                                  ["NarrowingType"] = "Double",        ["LargestVectorSize"] = "64", ["NextValueOp1"] = "Helpers.getMaskInt64()",                   ["NextValueOp2"] = "TestLibrary.Generator.GetUInt64()", ["NextValueOp3"] = "TestLibrary.Generator.GetDouble()"}),
```
[the remark this answers is not in the document]
Comment (SwapnilGaikwad) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs]: Done

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.cs (hunk 1 above)
Comment (SwapnilGaikwad) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.cs]: Done

**Comment by kunalspathak** [comment]
/ba-g superpmi replay failures is because of intermittent issue reported in https://github.com/dotnet/runtime/issues/103113 and other is related to https://github.com/dotnet/runtime/issues/104786

**Pull request merged**